### PR TITLE
[17.0][MIG] account_asset_management

### PR DESCRIPTION
**Pull request by sbiosca-s73**
Migration module account_asset_management to v17.0

### CODE REVIEW

**Review by hieulucky111 — APPROVED** [review]
We did some functional tests and it seems to work as expected.

File: account_asset_management/models/account_account.py [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+        help="Default Asset Profile when creating invoice lines with this account.",
+    )
+
+    @api.constrains("asset_profile_id")
+    def _check_asset_profile(self):
+        for account in self:
+            if (
+                account.asset_profile_id
+                and account.asset_profile_id.account_asset_id != account
+            ):
```
Comment (tien-ld) [inline account_asset_management/models/account_account.py:24]:
```suggestion
            if account.asset_profile_id.account_asset_id != account:
```

File: account_asset_management/models/account_asset_line.py [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+                                "prior to already posted entries."
+                            )
+                        )
+        return super().write(vals)
+
+    def unlink(self):
+        for dl in self:
+            if dl.type == "create" and dl.amount:
+                raise UserError(
+                    _("You cannot remove an asset line " "of type 'Depreciation Base'.")
```
Comment (tien-ld) [inline account_asset_management/models/account_asset_line.py]:
```suggestion
                    _("You cannot remove an asset line of type 'Depreciation Base'.")
```

File: account_asset_management/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            for aml in move.line_ids.filtered(
+                lambda line: line.asset_profile_id and not line.tax_line_id
+            ):
+                vals = move._prepare_asset_vals(aml)
+                if not aml.name:
+                    raise UserError(
+                        _("Asset name must be set in the label of the line.")
+                    )
+                if aml.asset_id:
+                    continue
```
Comment (tien-ld) [inline account_asset_management/models/account_move.py]:
```suggestion
                if not aml.name:
                    raise UserError(
                        _("Asset name must be set in the label of the line.")
                    )
                if aml.asset_id:
                    continue
                vals = move._prepare_asset_vals(aml)
```

File: account_asset_management/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        if "quantity" in vals or "asset_profile_id" in vals:
+            for record in self:
+                record._expand_asset_line()
+        return True
+
+    def _expand_asset_line(self):
+        self.ensure_one()
+        if self.asset_profile_id and self.quantity > 1.0:
+            profile = self.asset_profile_id
+            if profile.asset_product_item:
```
Comment (tien-ld) [inline account_asset_management/models/account_move.py]:
```suggestion
        if self.asset_profile_id.asset_product_item and self.quantity > 1.0:
```

**Comment by sbiosca-s73** [comment]
@tien-ld changes done!

File: account_asset_management/models/account_account.py (hunk 1 above)
Comment (sbiosca-s73) [reply account_asset_management/models/account_account.py:24]: @tien-ld  I think this change don't is correctly, because the condition `if account.asset_profile_id.account_asset_id` if account don't have asset_profile_id this always different of account. So, the condition must check asset_profile_id and later check this account_asset is different of account.

Comment (tien-ld) [reply account_asset_management/models/account_account.py:24]: ah, ok thank you

**Review by tien-ld — APPROVED** [review]

File: account_asset_management/views/account_asset.xml [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+                            <field
+                                name="name"
+                                class="oe_inline"
+                                readonly="state in ['open', 'close', 'removed']"
+                            />
+                        </h1>
+                    </div>
+                    <group>
+                        <group id="header_left_group">
+                            <field name="company_id" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                            <field name="company_id" invisible="1" />
```

File: account_asset_management/views/account_asset.xml [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+                            />
+                            <field
+                                name="company_id"
+                                widget="selection"
+                                groups="base.group_multi_company"
+                            />
+                            <field
+                                name="currency_id"
+                                groups="base.group_multi_currency"
+                                column_invisible="1"
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                                invisible="1"
```

File: account_asset_management/views/account_asset_profile.xml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                    />
+                    <div class="oe_title">
+                        <label for="name" class="oe_edit_only" />
+                        <h1>
+                                <field name="name" />
+                        </h1>
+                    </div>
+                    <group>
+                        <group>
+                            <field name="company_id" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset_profile.xml]:
```suggestion
                            <field name="company_id" invisible="1" />
```

File: account_asset_management/wizard/account_asset_remove.xml [hunk 8]
```diff
@@ -0,0 +1,58 @@
+<?xml version="1.0" encoding="utf-8" ?>
+<odoo>
+    <record id="account_asset_remove_view_form" model="ir.ui.view">
+        <field name="name">account.asset.remove.form</field>
+        <field name="model">account.asset.remove</field>
+        <field name="arch" type="xml">
+            <form string="Remove Asset">
+                <group>
+                    <group>
+                        <field name="company_id" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/wizard/account_asset_remove.xml]:
```suggestion
                        <field name="company_id" invisible="1" />
```

File: account_asset_management/views/account_move.xml [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                    optional="show"
+                />
+            </xpath>
+            <xpath
+                expr="//field[@name='invoice_line_ids']/form/sheet/group/field[@name='quantity']"
+                position="before"
+            >
+                <field
+                    name="asset_profile_id"
+                    column_invisible="parent.move_type not in ('in_invoice', 'in_refund')"
```
Comment (tien-ld) [inline account_asset_management/views/account_move.xml]:
```suggestion
                    invisible="parent.move_type not in ('in_invoice', 'in_refund')"
```

File: account_asset_management/views/account_move.xml [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+                />
+                <field
+                    name="asset_id"
+                    column_invisible="parent.move_type not in ('out_invoice', 'out_refund')"
```
Comment (tien-ld) [inline account_asset_management/views/account_move.xml]:
```suggestion
                    invisible="parent.move_type not in ('out_invoice', 'out_refund')"
```

File: account_asset_management/views/account_asset_profile.xml [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+                            <field name="group_ids" widget="many2many_tags" />
+                            <field name="asset_product_item" />
+                            <field name="active" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset_profile.xml]:
```suggestion
                            <field name="active" invisible="1" />
```

File: account_asset_management/views/account_asset_group.xml [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+<!-- pylint:disable=duplicate-xml-fields -->
+<odoo>
+    <record id="account_asset_group_view_form" model="ir.ui.view">
+        <field name="name">account.asset.group.form</field>
+        <field name="model">account.asset.group</field>
+        <field name="arch" type="xml">
+            <form string="Asset Group">
+                <group>
+                    <group>
+                        <field name="company_id" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset_group.xml]:
```suggestion
                        <field name="company_id" invisible="1" />
```

File: account_asset_management/views/account_asset.xml [hunk 13]
```diff
@@ -0,0 +1,422 @@
...
+                            />
+                            <field name="move_line_check" column_invisible="1" />
+                        </group>
+                        <group id="header_right_group">
+                            <field name="depreciation_base" />
+                            <field name="value_depreciated" />
+                            <field name="value_residual" />
+                            <field name="active" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                            <field name="active" invisible="1" />
```

File: account_asset_management/views/account_asset.xml [hunk 14]
```diff
@@ -0,0 +1,422 @@
...
+                                name="company_id"
+                                widget="selection"
+                                groups="base.group_multi_company"
+                            />
+                            <field
+                                name="currency_id"
+                                groups="base.group_multi_currency"
+                                column_invisible="1"
+                            />
+                            <field name="move_line_check" column_invisible="1" />
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                            <field name="move_line_check" invisible="1" />
```

File: account_asset_management/views/account_asset.xml [hunk 15]
```diff
@@ -0,0 +1,422 @@
...
+                                        groups="account.group_account_manager"
+                                        invisible="not move_check"
+                                    />
+                                </tree>
+                                <form string="Asset Line">
+                                    <group>
+                                        <group>
+                                            <field
+                                                name="parent_state"
+                                                column_invisible="1"
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                                                invisible="1"
```

File: account_asset_management/views/account_asset.xml [hunk 16]
```diff
@@ -0,0 +1,422 @@
...
+                                            />
+                                            <field
+                                                name="depreciation_base"
+                                                column_invisible="1"
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                                                invisible="1"
```

File: account_asset_management/views/account_asset.xml [hunk 17]
```diff
@@ -0,0 +1,422 @@
...
+                                            />
+                                            <field
+                                                name="currency_id"
+                                                column_invisible="1"
```
Comment (tien-ld) [inline account_asset_management/views/account_asset.xml]:
```suggestion
                                                invisible="1"
```

**Review by tien-ld — CHANGES_REQUESTED** [review]
Hi, `column_invisible` does not work on form view, please use `invisible`

**Comment by xtanuiha** [comment]
Hello, I have updated the Simplified Chinese translation `zh_CN.po`; please merge it, thank you!

[zh_CN.zip](https://github.com/OCA/account-financial-tools/files/14822086/zh_CN.zip)

**Comment by sbiosca-s73** [comment]
@tien-ld changes done. @xtanuiha I updated the translation

**Review by tien-ld — APPROVED** [review]

File: pandoc-3.1.11.1-1-amd64.deb
Comment (celm1990) [inline pandoc-3.1.11.1-1-amd64.deb:1]: Is this file needed?

**Review by celm1990 — COMMENTED** [review]
Please split the commits, separating changes from `pre-commit` into one commit and migration-related changes into another commit. Additionally, please squash commits related to administrative tasks, translations, etc. You can refer to https://github.com/OCA/maintainer-tools/wiki/Merge-commits-in-pull-requests#mergesquash-the-commits-generated-by-bots-or-weblate for guidance.

**Review by manuelregidor — APPROVED** [review]
Technical review. LGTM

File: account_asset_management/models/account_asset.py [hunk 18]
```diff
@@ -0,0 +1,1315 @@
...
+                    "period_amount": period_amount,
+                    "fy_amount": fy_amount,
+                    "day_amount": day_amount,
+                }
+            )
+            if self.method_time == "year":
+                fy_residual_amount -= fy_amount
+                if currency.is_zero(fy_residual_amount):
+                    break
+        i_max = i
```
Comment (xtanuiha) [inline account_asset_management/models/account_asset.py]: If the `table` variable is empty, it will not enter the loop; the variable `i` is uninitialized, and assigning it to `i_max` will trigger an error.
This will better:
```python
if table:
  i_max = i
``

**Comment by HaraldPanten** [comment]
@sbiosca-s73 will you continue with this PR?

THX.

**Comment by sbiosca-s73** [comment]
@celm1990 changes done, I split commits and deleted the unnecessary file.
@xtanuiha done.
@HaraldPanten

**Comment by HaraldPanten** [comment]
> @celm1990 changes done, I split commits and deleted the unnecessary file. @xtanuiha done. @HaraldPanten

THX! @sbiosca-s73 Could you check tests?

**Comment by sbiosca-s73** [comment]
@HaraldPanten tests checked

**Comment by HaraldPanten** [comment]
Could you separate pre-commit changes from migration commit?

THX.

**Comment by sbiosca-s73** [comment]
@HaraldPanten what changes?

**Comment by HaraldPanten** [comment]
When you open a new PR for a migration, you need to apply pre-commit, right? (black, isort, prettier).

This command makes some changes to your code and they have to be in a separate commit (not in the migration commit).

And then, you start the migration. These changes would be included in a new commit (the migration commit)

Here you can see several examples:

https://github.com/OCA/l10n-spain/pull/3583
https://github.com/OCA/account-invoice-reporting/pull/319

You'll see that the migration PRs that I mentioned have 2 commits. One of them is for the pre-commit stuff.

Maybe @ioans73 can explain you better

**Comment by pedrobaeza** [comment]
Following https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-17.0#technical-method-to-migrate-a-module-from-160-to-170-branch is what you obtain.

**Comment by ioans73** [comment]
@HaraldPanten @pedrobaeza Thanks for your comments :) I'm going to review it with @sbiosca-s73

**Comment by sbiosca-s73** [comment]
@HaraldPanten @pedrobaeza changes done, thanks!

**Review by HaraldPanten — APPROVED** [review]
Functional review. Overall, seems to be OK.

@pedrobaeza Could we merge this one?

File: account_asset_management/models/account_asset.py [hunk 19]
```diff
@@ -495,15 +474,16 @@ def name_search(self, name, args=None, operator="ilike", limit=100):
             if operator in expression.NEGATIVE_TERM_OPERATORS:
                 domain = ["&", "!"] + domain[1:]
         assets = self.search(domain + args, limit=limit)
-        return assets.name_get()
+        return assets._compute_display_name()
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: This is not correct. You can't call a compute method in a search.

File: account_asset_management/models/account_asset.py [hunk 20]
```diff
@@ -495,15 +474,16 @@ def name_search(self, name, args=None, operator="ilike", limit=100):
...
-    def name_get(self):
+    def _compute_display_name(self):
         result = []
         for asset in self:
             name = asset.name
             if asset.code:
                 name = " - ".join([asset.code, name])
+            asset.display_name = name
             result.append((asset.id, name))
         return result
```
Comment (pedrobaeza) [inline account_asset_management/models/account_asset.py]: A computed method shouldn't return anything. In fact, the computed method assign the value through dot notation.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
All the name search and display_name is wrong.

**Comment by sbiosca-s73** [comment]
@pedrobaeza changes done

File: account_asset_management/models/account_asset.py (hunk 19 above)
Comment (pedrobaeza) [reply account_asset_management/models/account_asset.py]: Instead of overriding `name_search`, use `_rec_names_search`, as stated in https://github.com/OCA/maintainer-tools/wiki/Migration-to-version-16.0 (it wasn't changed in previous migration)

Comment (sbiosca-s73) [reply account_asset_management/models/account_asset.py]: @pedrobaeza done

**Review by pedrobaeza — APPROVED** [review]
Thanks

/ocabot merge nobump
